### ISSUE TITLE
Update alert steals focus while typing in master password

### ISSUE THREAD

**Issue by kyldvs** [bitwarden/desktop#910]
## Describe the Bug

After launching the desktop app I will immediately begin typing in my master password. If there is an update, while I'm typing, the alert pops up and takes the focus away from the text field I was currently typing in. This often causes me to perform unintended actions without having a chance to even read the alert.

## Steps To Reproduce

1. Launch app that needs an update
2. Start typing in master password field
3. See update alert steal focus

## Expected Result

The update alert should not steal focus, not trigger while I'm typing, or wait until I've entered my master password.

## Actual Result

The update alert steal focus while I'm typing and causes unintended actions to be taken.

## Environment

- Operating system: Windows 10
- Installation method: Downloaded from Bitwarden.com
- Build Version: A version prior to 1.26.3, I don't know the exact version due to the bug itself, but it has been happening consistently for a long time

## Additional Context

I really like Bitwarden but this one issue is becoming increasingly frustrating every single time I launch the app. I would love some change to this alert to fix the issue

**Comment by thetillhoff** [comment]
Can confirm this happens with the upgrade popup from v1.26.4 to v1.26.5 on Windows 10.
This not only steals the focus while typing the master password but for all other applications as well.
So when you are running a full-screen application, this even minimizes that application, just to notify me bitwarden has an update available.
As I see it, bitwarden should use the win10 notifications like any other app, too. No reason to reinvent the wheel here...

This is nagging (me) for quite some time already, would be great to have this issue fixed as soon as possible - as its totally breaking the flow of whatever I'm doing at the time bitwarden shows that popup.

**Comment by parski** [comment]
This still happens as of the version before 1.30.0 and it also happens on Windows 11.

**Comment by kyldvs** [comment]
It would be great to get some sort of acknowledgment on this issue. It is the one and only frustrating aspect of Bitwarden that I run into almost every time I use Bitwarden. The typical case I run into this:

1. One of my sessions expired on a website, etc. and I need to login again. I'm actively working on logging into something so I can continue with what I'm doing.
2. I open Bitwarden
3. It opens and I start typing in my master password and press enter
4. Before I was able to hit enter, but after I started typing, an alert pops up asking me if I want to update Bitwarden
5. The alert gets focus and intercepts my enter from submitting my password (<-- this is the bug, I did not want this to happen, but have no time to react)
6. Bitwarden immediately closes and starts the update process
7. I'm now _very_ annoyed at Bitwarden because I was trying to do something else and there is no change that I will perceive from the update. I just want the update to happen silently on next launch or on close.

**Comment by eliykat** [comment]
Hey @kyldvs, sorry for the late reply. How do you think we could improve this? We probably don't want to wait until the user has logged in before offering to update. Maybe a debounce to wait X seconds after the user has last changed the value of the login form?

**Comment by parski** [comment]
Either wait to prompt until X seconds after last input or maybe defer the prompt until next application launch.

Well funded applications might display a small banner at the top of the window during the session if an update is available (and/or downloaded) but I think either aforementioned behavior is good enough to make the user experience **much** better.

**Comment by kyldvs** [comment]
My ideal implementation would be:

Updates are split between critical/non-critical. Critical updates can maintain the same alerting behavior that steals focus, they are critical, I don't mind if you throw them in my face as much. Non-critical updates on the other hand don't alert until you have logged in and just display as a banner at the top, or some badge in the interface indicating that an update is available, but never as an alert that takes focus.

---

That seems like a fair bit of work, so your solution of waiting for X seconds of inactivity before alerting would be a great first step. I still believe the alert that would pop up once I'm logged in will still be annoying, but much less so since it will be less likely to intercept an "enter" press.

One situation I could see with this change that would make the entire situation _worse_ would be if the alert makes the window grab focus at the operating system level. I don't recall if that is the case, but if I'm in another application and the now debounced alert makes bitwarden grab focus from my current application that would be _worse_ than the current behavior.

**Comment by thetillhoff** [comment]
I think the solution is even simpler:
Don't make the check whether an update is available while the password prompt is visible. Do it before that (app-launch-screen).

I (and probably most people) watch my screen when I open an app to see if I have the expected input prompt and start typing as soon as I find it.
What is then disturbing me is the lose of focus midway. This means the check is currently happening in parallel - while I prefer the sequential approach in this case. And since the update might render the input 'useless' I prefer the update check before the password prompt.
For performance reasons, this could be done at app start only, not every time it's opened.

If needed:
Afterwards, the check can happen periodically or when the ui is opened, and then a banner can appear - this can happen in parallel, since it's not as intrusive and hinders workflows.

**Comment by eliykat** [comment]
Thanks for your feedback, all very useful. I've raised it with our design team and I'll see what we can do to improve the autoupdate process.

**Comment by kyldvs** [comment]
Another issue I've noticed is that the "accept" option is automatically focused and "space" triggers the focused item in the alert. It seems to me the "do nothing" option should be the default focus in most alerts, and they should only trigger on "enter" (but perhaps the keys that trigger it is a native windows thing).
